src/omniglot/wrapper.py

```diff
@@ -139,8 +140,182 @@ def run_batches(self, batches, optimizer, train=False, meta_train=False):
+            if 'momentum' in optimizer_kwargs else 0.)
+
+        # For now it is a dummy updater does nothing in backward call.
+        updater = warpgrad.SimpleUpdater(criterion, **meta_kwargs)
+
+        # we don't need replay buffer for algorithm1
+        model = warpgrad.Warp(model=model,
+                              adapt_modules=list(model.adapt_modules()),
+                              warp_modules=list(model.warp_modules()),
+                              updater=updater,
+                              buffer=None,
+                              optimizer_parameters=optimizer_parameters)
+
+        super(WarpGradOnlineWrapper, self).__init__(criterion,
+                                                    model,
+                                                    optimizer_cls,
+                                                    optimizer_kwargs)
 
+        self.meta_optimizer_cls = optim.SGD \
+            if meta_optimizer_cls.lower() == 'sgd' else optim.Adam
+        lra = meta_optimizer_kwargs.pop(
+            'lr_adapt', meta_optimizer_kwargs['lr'])
+        lri = meta_optimizer_kwargs.pop(
+            'lr_init', meta_optimizer_kwargs['lr'])
+        lrl = meta_optimizer_kwargs.pop(
+            'lr_lr', meta_optimizer_kwargs['lr'])
+        self.meta_optimizer = self.meta_optimizer_cls(
+            [{'params': self.model.init_parameters(), 'lr': lri},
+             {'params': self.model.warp_parameters(), 'lr': lra},
+             {'params': self.model.optimizer_parameters(), 'lr': lrl}],
+            **meta_optimizer_kwargs)
+
+        # This is the meta loss that we are going to accumulate.
+        self.meta_loss = 0
+
+    def _partial_meta_update(self, loss, final):
+        pass
+
+    def _final_meta_update(self):
+
+        def step_fn():
+            self.meta_optimizer.step()
+            self.meta_optimizer.zero_grad()
+
+        self.model.backward(step_fn, **self.optimizer_kwargs)
+
+    def run_tasks(self, tasks, meta_train):
+        """Train on a mini-batch tasks and evaluate test performance.
+
+        Arguments:
+            tasks (list, torch.utils.data.DataLoader): list of task-specific
+                dataloaders.
+            meta_train (bool): whether current run in during meta-training.
+        """
+        results = []
+        self.meta_loss = 0
+        for task in tasks:
+            task.dataset.train()
+            trainres = self.run_task(task, train=True, meta_train=meta_train)
+            task.dataset.eval()
+            valres = self.run_task(task, train=False, meta_train=False)
+            results.append((trainres, valres))
+        ##
+        results = AggRes(results)
+
+        # Meta gradient step
+        if meta_train:
+            # at the end of collection for K steps N tasks we do the backward
+            # pass.
+            backward(self.meta_loss, self.model.meta_parameters(
+                include_init=False))
+            self._final_meta_update()
```

Comment:
When we collect `k` times inner iteration for `N` tasks we can call backward pass calculate gradients.